src/main/java/htsjdk/samtools/HtsgetBAMFileReader.java

```diff
@@ -0,0 +1,561 @@
+        }
+
+        @Override
+        public SAMRecord next() {
+            final SAMRecord result = this.currentRecord;
+            this.advance();
+            return result;
+        }
+
+        private void advance() {
+            while (this.currentIterator != null && !this.currentIterator.hasNext()) {
+                this.advanceIterator();
+            }
+            this.currentRecord = this.currentIterator == null ? null : this.currentIterator.next();
+        }
+
+        private void advanceIterator() {
+            if (this.currentIterator != null) {
+                this.currentIterator.close();
+            }
+            if (!this.iterators.hasNext()) {
+                this.currentIterator = null;
+                return;
+            }
+            final Interval currInterval = this.intervals[this.currentIntervalIndex];
+            final Interval prevInterval = this.currentIntervalIndex == 0 ? null : this.intervals[this.currentIntervalIndex - 1];
+            final SamRecordFilter filter = new SamRecordFilter() {
+                @Override
+                public boolean filterOut(final SAMRecord record) {
+                    return record.getReadUnmappedFlag() && record.getAlignmentStart() != SAMRecord.NO_ALIGNMENT_START
+                        ? !this.acceptUnmappedRecord(record)
+                        : !this.acceptRecord(record);
+                }
+
+                @Override
+                public boolean filterOut(final SAMRecord first, final SAMRecord second) {
+                    throw new UnsupportedOperationException();
+                }
+
+                private boolean acceptRecord(final SAMRecord rec) {
+                    return BAMQueryChainingIterator.this.contained
+                        ? currInterval.contains(rec) && (prevInterval == null || !prevInterval.contains(rec))
+                        : currInterval.overlaps(rec) && (prevInterval == null || !prevInterval.overlaps(rec));
+                }
+
+                private boolean acceptUnmappedRecord(final SAMRecord rec) {
+                    final int start = rec.getStart();
+                    final boolean matchesCurrInterval =
+                        rec.contigsMatch(currInterval) &&
+                            CoordMath.encloses(currInterval.getStart(), currInterval.getEnd(), start, start);
+                    final boolean matchesPrevInterval =
+                        prevInterval != null &&
+                            rec.contigsMatch(prevInterval) &&
+                            CoordMath.encloses(prevInterval.getStart(), prevInterval.getEnd(), start, start);
+                    return matchesCurrInterval && !matchesPrevInterval;
+                }
+            };
+            this.currentIterator = new FilteringSamIterator(this.iterators.next().get(), filter, false);
+            this.currentIntervalIndex++;
+        }
+    }
+
+    public static class BAMQueryFilteringIterator implements CloseableIterator<SAMRecord> {
```

Comment:
It's a shame there isn't a good way to avoid duplicating this code identically from BamReader